src/design_notes/dynsized_constraints.md

```diff
@@ -0,0 +1,140 @@
+# Exotically sized types (`DynSized` and `extern type`)
+
+## Overview
+
+In current Rust, there's two kinds of types with respect to sizing:
+if a type is `Sized`, its layout (size and alignment) is known statically,
+and if a type is `?Sized`, its layout may not be known until runtime (e.g. via a vtable).
+
+However, more exotically sized types exist; the most common example is opaque `extern type`.
+`extern type`s have an *unknown* layout to Rust, and as such can only be used behind a pointer type.
+Since the most unsized a type can currently be is `?Sized`, though,
+the compiler has to make up a size and alignment to return from `mem::size_of_val`/`align_of_val`.
+Currently the compiler returns a size of 0 and an alignment of 1.
+Lying in this fashion is considered undesirable \[2].
+
+Additionally, some C-header-interface libraries expose an opaque (incomplete) type
+but also provide a function returning the size of the type and expect the caller to allocate space.
+This is useful to allow the library to change the size of the type,
+but still allow the caller to control allocation (e.g. using a custom arena allocator).
+When bridging to Rust, these types should ideally have access to dynamic size/align.
+
+## Proposed Solution
+
+The most obvious and independently reinvented solution is a "`DynSized`" trait that provides dynamic size/align information.
+`extern type` would not implement `DynSized`, and generic code could opt into `?DynSized` types to support such.
+
+At the time of writing, there is weak approval from T-lang to proceed with an internal-only version of `DynSized`
+which is used to prohibit the use of `extern type` in standard `<T: ?Sized>` generic arguments \[2].
+
+This design document is about the restrictions on what `T: ?Sized + DynSized` actually needs to imply.
+
+## Design Constraints
+
+### `Arc` and `Weak`
+
+`Arc` supports "zombie" references, where all strong `Arc` and the pointee have been dropped,
+but `Weak` handles still exist and so the allocation still exists.
+This means that `Weak` needs to be able to determine the layout of the allocation from a dropped pointee.
+
+In addition, `Weak` are pointers to the *reference count* part of the `ArcInner` allocation,
+and thus need to *statically* know the alignment of the pointee type to determine the offset
+(it cannot call `align_of_val_raw` without first knowing the offset).
+
+For the alignment, there are three potential resolutions:
+
+- Store layout information in the `ArcInner` header,
+- Require that alignment be determined solely from pointee metadata, or
+- Change the pointer of `Arc<T>` to point directly at `T` and use a fixed negative offset for the header.
+
+For the both, there are three potential resolutions:
+
+- Store layout information in the `ArcInner` header, or
+- Require that layout be determined solely from pointee metadata, or
+- Require that layout be determinable from a dropped pointee.
```

Comment:
I have no idea how this could be done -- are these bullets just a duplication of the bullets above?

Reply:
Reading on, I guess I understand, I feel like this could be clarified:

```suggestion
- Require that layout be determinable from a dropped pointee. This is needed because a `Weak` ref holds a reference to the allocation, even when the content may have been dropped. One way to achieve this is to have the layout be stored in the metadata of the wide pointer; another would be to ensure that drop does not overwrite the fields storing the layout information (e.g., a thin pointer with an inline vtable would still have the vtable even if the contents are dropped).
```

Reply:
I restructured that section a bit to hopefully make this clearer before reading the rest, and put some of this context in a footnote; hopefully that should be sufficient.

In theory, the above paragraph should be what illustrates the need for determining layout after dropping `T`.